src/bls12_381/fq.rs

```diff
@@ -804,6 +821,33 @@ impl Fq {
         (self.0).0[5] = r11;
         self.reduce();
     }
+
+    /// Hash into the field.
+    /// Hashing consists of BLAKE2b in counter mode, truncated and shaved bits,
+    /// then interpreted it (in big endian) as Fr/Fq elements.
+    pub fn hash(k: &[u8], nonce: &[u8]) -> Self {
+        // the key must be of at least 128 bits.
+        assert!(k.len() >= 16);
+
+        let mut repr: [u64; 6] = [0; 6];
+        let mut count : u32 = 0;
+        let mut count_u8 : [u8; 4] = [0; 4];
+
+        loop {
+            // increment the counter
+            count += 1;
+            NativeEndian::write_u32(&mut count_u8, count);
+            // hash in counter mode: append to the nonce a counter
+            let mut hasher = Blake2b::new_keyed(k);
+            hasher.input(nonce);
+            hasher.input(&count_u8);
+            // truncate and shave the hash result
+            BigEndian::read_u64_into(&hasher.result().as_slice()[.. 48], &mut repr);
+            let mut e = Fq(FqRepr(repr));
+            e.0.divn(REPR_SHAVE_BITS);
+            if e.is_valid() { return e }
+        }
+    }
 }
 
 impl SqrtField for Fq {
```

Comment:
Shouldn't legendre implement blinding by a factor `r^2` for a random `r \in F_q` as suggested in the Fouque-Tibouchi paper? Or is timing security of the quadratic character check not significant enough?

Reply:
From the `README`
```
This library does not make any guarantees about constant-time operations, memory access patterns, or resistance to side-channel attacks.
```
for more information, you can also check my commit messages, Foque's comment, or the implementation of double-and-add of this library.

Reply:
@ValarDragon If you go with @ebfull suggestion, no legendre is computed *ever*. With blinding, you instead introduce random factor which really doesn't need to be here.